contracts/oracle/src/contract/alarms/mod.rs

```diff
@@ -10,101 +12,142 @@ use marketprice::{alarms::PriceAlarms, SpotPrice};
 use sdk::cosmwasm_std::{Addr, Storage};
 use swap::SwapGroup;
 
-use crate::{alarms::Alarm as AlarmDTO, ContractError};
+use crate::{alarms::Alarm as AlarmDTO, error::ContractError, result::ContractResult};
 
-use self::iter::AlarmsFlatten;
+use self::iter::Iter as AlarmsIter;
 
 mod iter;
 
-pub type PriceResult<BaseC> = Result<BasePrice<SwapGroup, BaseC>, ContractError>;
+pub type PriceResult<BaseC> = ContractResult<BasePrice<SwapGroup, BaseC>>;
+
+const NAMESPACE_ALARMS_BELOW: &str = "alarms_below";
+const NAMESPACE_INDEX_BELOW: &str = "index_below";
+const NAMESPACE_ALARMS_ABOVE: &str = "alarms_above";
+const NAMESPACE_INDEX_ABOVE: &str = "index_above";
+const NAMESPACE_IN_DELIVERY: &str = "in_delivery";
 
-pub struct MarketAlarms;
+pub(super) struct MarketAlarms<'storage, S>
+where
+    S: Deref<Target = dyn Storage + 'storage>,
+{
+    alarms: PriceAlarms<'storage, S>,
+}
 
-impl MarketAlarms {
-    const PRICE_ALARMS: PriceAlarms<'static> =
-        PriceAlarms::new("alarms_below", "index_below", "alarms_above", "index_above");
+impl<'storage, S> MarketAlarms<'storage, S>
+where
+    S: Deref<Target = dyn Storage + 'storage>,
+{
+    pub fn new(storage: S) -> Self {
+        Self {
+            alarms: PriceAlarms::new(
+                storage,
+                NAMESPACE_ALARMS_BELOW,
+                NAMESPACE_INDEX_BELOW,
+                NAMESPACE_ALARMS_ABOVE,
+                NAMESPACE_INDEX_ABOVE,
+                NAMESPACE_IN_DELIVERY,
+            ),
+        }
+    }
+
+    pub fn notify_alarms_iter<I, BaseC>(&self, prices: I) -> AlarmsIter<'storage, '_, S, I, BaseC>
```

Comment:
I am not able to do so unless I add a bound which states that the iterator type `I` should outlive the lifetime `'storage`.
That is because under the hood, I capture both lifetimes which are on their own independent.
Another thing is, I don't think that it's necessarily bad to state the type as long as it doesn't imply what the implementation is, unlike the `iter::Map<I, ...>` types that I used in previous iterations.